root/roofit/roofit-ad/RooFuncWrapperBenchmarks.cxx

```diff
@@ -0,0 +1,122 @@
+   RooAddition muShifted("mu_shifted", "mu_shifted", {mu, shift});
+
+   RooRealVar sigma("sigma", "sigma", 4.0, 0.01, 10);
+   RooConstVar scale("scale", "scale", 1.5);
+   RooProduct sigmaScaled("sigma_scaled", "sigma_scaled", sigma, scale);
+
+   RooGaussian gauss{"gauss", "gauss", x, muShifted, sigmaScaled};
+
+   RooArgSet normSet{x};
+
+   // Generate the same dataset for all backends.
+   RooRandom::randomGenerator()->SetSeed(100);
+
+   std::size_t nEvents = state.range(1);
+   std::unique_ptr<RooDataSet> data{gauss.generate(x, nEvents)};
+
+   // Set the values away from the minimum
+   mu.setVal(-2.0);
+   sigma.setVal(7.0);
+
+   // Save the original parameter values and errors to reset after minimization
+   RooArgSet params;
+   gauss.getParameters(&normSet, params);
+   RooArgSet origParams;
+   params.snapshot(origParams);
+
+   std::unique_ptr<RooAbsReal> nllRef{gauss.createNLL(*data, RooFit::BatchMode("off"))};
+   std::unique_ptr<RooAbsReal> nllRefBatch{gauss.createNLL(*data, RooFit::BatchMode("cpu"))};
+   auto nllRefResolved = static_cast<RooAbsReal *>(nllRefBatch->servers()[0]);
+
+   std::string name = "myNll" + std::to_string(counter);
+   RooFuncWrapper nllFunc(name.c_str(), name.c_str(), *nllRefResolved, normSet, data.get());
+
+   std::unique_ptr<RooMinimizer> m = nullptr;
+   for (auto _ : state) {
+      int code = state.range(0);
+      if (code == 0) {
+         m.reset(new RooMinimizer(*nllRef));
+      } else if (code == 1) {
+         m.reset(new RooMinimizer(nllFunc));
+      } else if (code == 2) {
+         m.reset(new RooMinimizer(*nllRefBatch));
+      } else {
+         RooMinimizer::Config minimizerCfgAd;
+         minimizerCfgAd.gradFunc = [&](double *out) { nllFunc.getGradient(out); };
+         m.reset(new RooMinimizer(nllFunc, minimizerCfgAd));
+      }
+
+      m->setPrintLevel(-1);
+      m->setStrategy(0);
+      params.assign(origParams);
+      m->minimize("Minuit2");
+   }
+}
+
+auto const unit = benchmark::kMillisecond;
+
+// Run the minimization with the reference NLL
+BENCHMARK(BM_RooFuncWrapper_Minimization)
+   ->Name("NllReferenceMinimization")
+   ->ArgsProduct({{0}, benchmark::CreateRange(10, 10000, 10)})
+   ->Unit(unit)
+   ->Iterations(100);
+
+// Run the minimization with the reference NLL (BatchMode)
+BENCHMARK(BM_RooFuncWrapper_Minimization)
+   ->Name("NllBatchModeMinimization")
+   ->ArgsProduct({{3}, benchmark::CreateRange(10, 10000, 10)})
+   ->Unit(unit)
+   ->Iterations(100);
+
+// Run the minimization with the code-squashed version with numerical-diff.
+BENCHMARK(BM_RooFuncWrapper_Minimization)
+   ->Name("NllCodeSquash_NumDiff")
+   ->ArgsProduct({{1}, benchmark::CreateRange(10, 10000, 10)})
+   ->Unit(unit)
+   ->Iterations(100);
+
+// Run the minimization with the code-squashed version with AD.
+BENCHMARK(BM_RooFuncWrapper_Minimization)
+   ->Name("NllCodeSquash_AD")
+   ->ArgsProduct({{2}, benchmark::CreateRange(10, 10000, 10)})
+   ->Unit(unit)
+   ->Iterations(100);
```

Comment:
I expect all of the subsequent benchmarks to be comparing against these 4 elements. Can we create some benchmark fixture for these types of benchmarks?

Reply:
I dont think a benchmark fixture would make sense here but I can move them into a function?

Reply:
@vgvassilev I have added a utility function to run all the benchmarks in the new commit.